clang/test/CodeGenSYCL/code_location.cpp

```diff
@@ -0,0 +1,266 @@
+// RUN: %clang_cc1 -fsycl-is-device -internal-isystem -sycl-std=2020 -fsycl-int-header=%t.h %s -o %t.out
+// RUN: FileCheck -input-file=%t.h %s
+// RUN: %clang_cc1 -fsycl-is-host -sycl-std=2020 %s | FileCheck -input-file=%t.h %s
+
+#include "Inputs/sycl.hpp"
+
+// Check that meaningful information is returned when NDEBUG is not defined
+// and empty strings and 0s are emitted when it is.
+int test1() {
+  cl::sycl::queue q;
+  q.submit([&](cl::sycl::handler &h) { h.single_task([] {}); });
+  q.submit([&](cl::sycl::handler &h) { h.single_task<class KernelName>([]() {}); });
+  return 0;
+}
+// CHECK: template <> struct KernelInfoData<'_', 'Z', 'T', 'S', 'Z', 'Z', '5', 't', 'e', 's', 't', '1', 'v', 'E', 'N', 'K', 'U', 'l', 'R', 'N', '2', 'c', 'l', '4', 's', 'y', 'c', 'l', '7', 'h', 'a', 'n', 'd', 'l', 'e', 'r', 'E', 'E', '_', 'c', 'l', 'E', 'S', '2', '_', 'E', 'U', 'l', 'v', 'E', '_'> {
+// CHECK:   static constexpr const char* getFileName() {
+// CHECK: #ifndef NDEBUG
+// CHECK:     return "code_location.cpp";
+// CHECK: #else
+// CHECK:     return "";
+// CHECK: #endif
+// CHECK:   }
+// CHECK:   static constexpr const char* getFunctionName() {
+// CHECK: #ifndef NDEBUG
+// CHECK:     return "";
```

Comment:
Is this a sufficient result for this test case?

Reply:
There is no enclosing namespace or anything akin to that in this test case. -Since it is an unnamed lambda, the empty string for functionname, combined with its location and filename information seems to make sense, doesn't it? @alexbatashev

Reply:
Yes, I would like to hear @alexbatashev's thoughts on this one and the one before.   If the requirements are believed to be satisfied, I am fine with it.

Reply:
That's a tricky question. On one hand, in both SPIR-V and profilers we will see some definite kernel name, on the other we won't be able to do much with that name, given some unusual mangling rules for these lambdas. I suggest we leave it as-is, we'll be able to construct hashes with just file name, line and column, and we can add something here in future if it is needed.

Reply:
Thanks @alexbatashev